Applications/Spire/Source/Ui/AdaptiveBox.cpp

```diff
@@ -0,0 +1,415 @@
+
+        bool eventFilter(QObject* watched, QEvent* event) override {
+          if(event->type() == QEvent::Show ||
+              event->type() == QEvent::ShowToParent) {
+            m_item->set_visible(true);
+          } else if(event->type() == QEvent::Hide ||
+              event->type() == QEvent::HideToParent) {
+            m_item->set_visible(false);
+          }
+          return QObject::eventFilter(watched, event);
+        }
+      };
+
+      WidgetItem m_item;
+      std::unique_ptr<EventFilter> m_filter;
+      std::shared_ptr<SharedWidget> m_shared_widget;
+      bool m_is_visible;
+      QWidget* m_parent;
+
+      void take_control() {
+        if(is_visible()) {
+          if(m_shared_widget->take_control(this)) {
+            m_item.widget()->setParent(m_parent);
+            m_item.widget()->show();
+          }
+        } else if(m_shared_widget->release_control(this)) {
+          m_item.widget()->setParent(nullptr);
+        }
+      }
+  };
+
+  struct GridLayoutItem {
+    QLayoutItem* m_item;
+    int m_row;
+    int m_column;
+    int m_row_span;
+    int m_column_span;
+  };
+
+  struct BoxLayoutItem {
+    QLayoutItem* m_item;
+    int m_stretch;
+  };
+
+  struct LayoutItem {
+    QLayoutItem* m_item;
+  };
+
+  auto make_placeholder(QWidget& widget, QWidget& parent,
+      const QLayoutItem& item) {
+    auto placeholder = new PlaceholderLayoutItem(widget);
+    placeholder->set_parent(parent);
+    placeholder->setAlignment(item.alignment());
+    return placeholder;
+  }
+
+  template<typename L, typename T>
+  void replace(L& layout, QWidget& parent,
+      std::function<T(L& layout)> remove,
+      std::function<void(L& layout, T& item, QWidget& widget)> add_widget,
+      std::function<void(L& layout, T& item, QLayout& child_layout)> add_layout,
+      std::function<void(L& layout, T& item)> add_others) {
+    auto items = std::queue<T>();
+    while(layout.count() > 0) {
+      items.push(remove(layout));
+    }
+    while(!items.empty()) {
+      auto& item = items.front();
+      items.pop();
+      if(auto widget = item.m_item->widget()) {
+        add_widget(layout, item, *widget);
+      } else if(auto child_layout = item.m_item->layout()) {
+        add_layout(layout, item, *child_layout);
+      } else {
+        add_others(layout, item);
+      }
+    }
+  }
+
+  void replace_widget_with_placeholder(QLayout& layout, QWidget& parent) {
+    if(auto grid_layout = dynamic_cast<QGridLayout*>(&layout)) {
+      replace<QGridLayout, GridLayoutItem>(*grid_layout, parent,
+        [] (QGridLayout& layout) {
+          auto item = GridLayoutItem();
+          layout.getItemPosition(0, &item.m_row, &item.m_column,
+            &item.m_row_span, &item.m_column_span);
+          item.m_item = layout.takeAt(0);
+          return item;
+        },
+        [&] (QGridLayout& layout, GridLayoutItem& item, QWidget& widget) {
+          layout.addItem(make_placeholder(widget, parent, *item.m_item),
```

Comment:
What is being captured?

Reply:
`parent` is being captured by the lambda expression.